### PR TITLE
Adding a dlrn sub-module

### PR DESCRIPTION
**Pull request by fuzzball81**
This is the start of a dlrn module the will start to drive atkinson.
This portion handles a few of the basic http requests that we have to
make to extract version data and what the current commit is for a given
promoted link.

We now rely on toolchest version 0.0.4 or greater for some of the
functionality.

Signed-off-by: Jason Joyce <fuzzball81@gmail.com>

### CODE REVIEW

File: atkinson/dlrn/http_data.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        :param url: The URL to the host to obtain data.
+        :param releases: The release name to use for lookup.
+        """
+        self.url = url
+        self.release = release
+
+    def _build_url(self, commit_hash, distgit_hash):
+        """
+        Generate a url given a commit hash and distgit hash to match the format
+        base/AB/CD/ABCD123_XYZ987 where ABC123 is the commit hash and XYZ987
```
Comment (jguiditta) [inline atkinson/dlrn/http_data.py]: ABCD123 ^

File: atkinson/dlrn/http_data.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        is a portion of the distgit hash.
+
+        :param commit_hash: A string with the full commit hash on the change
+        :param dlrn_hash: A string with the full dlrn hash to fetch from
+        :return: A string with the full URL.
+        """
+        first = commit_hash[0:2]
+        second = commit_hash[2:4]
```
Comment (jguiditta) [inline atkinson/dlrn/http_data.py:102]: Just to be sure, dlrn _always_ uses the first 4 digits of the commit hash and splits them in two to generate to url?

File: atkinson/dlrn/http_data.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        ret_dict = {}
+        full_url = os.path.join(self._build_url(commit_hash, distgit_hash),
+                                'versions.csv')
+        data = _raw_fetch(full_url)
+        if data is not None:
+            data = data.replace(' ', '_')
+            split_data = data.split()
+            reader = csv.DictReader(split_data)
+            for row in reader:
+                print(row)
```
Comment (jguiditta) [inline atkinson/dlrn/http_data.py]: is this a debug line?

File: tests/unit/dlrn/test_http_data.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            dlrn = dlrn_http_factory('tester', config_file=config_file)
+            actual = dlrn.commit('consistent')
+            assert expected == actual
+
+
+def test_good_data_bad_status(datadir):
+    """
+    GIVEN we have a DlrnHttpData instance
+    WHEN we fetch the commit yaml and it contains data
+    AND the status code is good
```
Comment (jguiditta) [inline tests/unit/dlrn/test_http_data.py]: status code is bad? ^

File: atkinson/dlrn/http_data.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                                self.release,
+                                name, 'commit.yaml')
+        data = _fetch_yaml(full_url)
+        if data is not None and 'commits' in data:
+            for pkg in data['commits']:
+                if pkg['status'] == 'SUCCESS':
+                    ret_data.append({'name': pkg['project_name'],
+                                     'dist_hash': pkg['distro_hash'],
+                                     'commit_hash': pkg['commit_hash']})
+
```
Comment (jguiditta) [inline atkinson/dlrn/http_data.py:130]: I wonder if perhaps we should log bad status or something, since the method returns an empty list in that case?

**Review by jguiditta — COMMENTED** [review]
Looks good overall, couple comments/questions

File: atkinson/dlrn/http_data.py (hunk 2 above)
Comment (fuzzball81) [reply atkinson/dlrn/http_data.py:102]: Yes, that is format that dlrn uses to build out its folder structure. For example: https://trunk.rdoproject.org/centos7-master/60/5a/605ad27f053a1ef9ffb3601c6023987ddb6a22e1_44bb75ad/

File: atkinson/dlrn/http_data.py (hunk 1 above)
Comment (fuzzball81) [reply atkinson/dlrn/http_data.py]: Good catch

File: atkinson/dlrn/http_data.py (hunk 3 above)
Comment (fuzzball81) [reply atkinson/dlrn/http_data.py]: Yes, I forgot about that one.

File: tests/unit/dlrn/test_http_data.py (hunk 4 above)
Comment (fuzzball81) [reply tests/unit/dlrn/test_http_data.py]: Bad copy paste.

File: atkinson/dlrn/http_data.py (hunk 5 above)
Comment (fuzzball81) [reply atkinson/dlrn/http_data.py:130]: Yes, that was something I was going to add. When I started some of the development we did not have the logger module merged back. I will add some logging here and in a few other places.

**Review by jguiditta — APPROVED** [review]
Looks good! :+1:

**Pull request merged**